rfcs/0076-refined-s3-source.md

```diff
@@ -0,0 +1,88 @@
+  too many files.
+
+# Solution
+
+The intuitive solution is to use state store to store the file list, and it is kind of a good idea.
+But it is not easy for Meta to access Hummock, the storage engine, and even if it could write the file list down to state
+store, it is even harder to align the data with barrier, which means it can be dangerous for source executors to use the list.
+
+## List and Fetch
+
+So I am proposing a new solution, which is to use two separate executors to complete the job.
+Let's say they are `list` and `fetch`, aligning the name in [RFC: S3 Source with SQS](https://github.com/risingwavelabs/rfcs/pull/22), 
+and they do the same thing.
+
+* `List` is unique in one source plan. It is responsible for keep listing all files and push the filenames to the downstream.
+* `Fetch` is responsible for fetching the file from S3 and push the file to the downstream, as well as write the received
+  filenames to Hummock.
+
+## How It Works
+
+![img](../assets/refined-s3-source.svg)
+
+### List Part
+
+The `list` cannot push all filenames to the `fetch` part every time. Letting the shuffle operator distribute millions of
+rows every minute is an excessive and unnecessary burden on the system.
+If we do this, the fetch operator must perform `point get` operation on each row to deduplicate, which is also unacceptable for hummock.
+
+The `list` executor will keep listing the files in S3 bucket and try its best to de-duplicate the files (but no guarantee).
+It will record the timestamp when doing the last listing and filter out the files that are created before the timestamp.
+And we can utilize the idea of [RFC: Refine S3 Source: List Objects in A Lazy Way](https://github.com/risingwavelabs/rfcs/pull/74),
+not to list all pages at once, but to list the pages one by one, preventing overload downstream.
+
+`list` executor will keep the max page `p` number as well as the last list timestamp `ts` in its state store.
+When list op is triggered, it filters out the files that are created before `ts` in the first `p` pages.
+If `p` increases (`list` only increase `p` at the request of Meta by barrier), it will push down all items in the `p+1` page
+and update `p` and `ts` in the state store.
+
+### Fetch Part
+
+The `fetch` part will do the ordinary thing, get files by name, read, and yield stream chunks.
+Each `fetch` executor accepts both message stream (filenames from `list`) and barrier stream.
+
+Every `fetch` executor keeps a local task queue, and a state cache in the memory.
+
+* Task queue is used to store the filenames that are not fetched yet.
+  * Task queue is initialized by `select * from state_table where partition in local_vnode`. It just reads all things in the local vnode.
+  * When receiving new filenames from the message stream, it will push the filenames to the task queue.
+* State cache is used to store the filenames that are fetched but not written to hummock yet.
```

Comment:
I have two questions here:
1. How many state tables are maintained in Fech executor? What are the schemas?
2. What does the state transition look like for each file?

Here are my thoughts:
- Two state tables "Task Queue" and "File State" are maintained. The former is the partial index of the later one, meaning that files in "Task Queue" must be present in "File State". Schemas:
```
"Task Queue"
| file_name | read_offset | enqueue_epoch? (can be used to sort tasks on recovery) |

"File State"
| file_name (bloom_filter_key) | metadata (size/enqueue_epoch/finish_epoch/...) | 
```
- State transitions for a file (these are logical stated and may not all exist in the implementation)
  - `INIT`: file name is received from list executor or is seen after reading from state store on recovery
  - `INIT` -> `FINISH`: file name is present in "File State" and absent in "Task Queue"
  - `INIT` -> `IN_PRORESS`: file name is absent in "File State". **Insert into "File State" and "Task Queue".**
  - `IN_PROGRESS` -> `IN_PROGRESS`: read_offset updated. **Update read_offset "Task Queue"**.
  - `IN_PROGRESS` -> `FINISH`: finish reading the file. **Delete from "Task Queue" and Update finish_epoch in "File State"**

Reply:
for `fetch`, we just need one state table. Here is the schema. 

```
| filename | read offset | status | 
```

In my design, the task queue is in memory.
Every time we insert a new file, it will be `(filename, NULL, INIT)` and for reading file, the row will be `(filename, offset, GOING | DONE)`

I think we don't need to maintain the order of each file in the task queue.

Reply:
The main benefit of having a persistent state of received but unfinished files is that we don't need to do a full table scan on the state table on recovery. In other words, this extra state table is an index. More files will be inserted into the file state table and the states never be cleaned up. Therefore, I exepct that recovery will become slower over time without the extra index.

The order is just a side-benefit of persisting the task queue and I agree the order doesn't affect correctness.

Reply:
Oh, I underestimated the overhead of doing a table scan. Then, your solution fits the scene best for you. We can keep the task queue at the thousand-row level with pagination mechanism.

Reply:
> Oh, I underestimated the overhead of doing a table scan. Then, your solution fits the scene best for you. We can keep the task queue at the thousand-row level with pagination mechanism.

SGTM.